pkg/flushqueues/exclusivequeues.go

```diff
@@ -0,0 +1,64 @@
+package flushqueues
+
+import (
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/uber-go/atomic"
+)
+
+type ExclusiveQueues struct {
+	queues     []*util.PriorityQueue
+	index      *atomic.Int32
+	activeKeys sync.Map
+}
+
+// New creates a new set of flush queues with a prom gauge to track current depth
+func New(queues int, metric prometheus.Gauge) *ExclusiveQueues {
+	f := &ExclusiveQueues{
+		queues: make([]*util.PriorityQueue, queues),
+		index:  atomic.NewInt32(0),
+	}
+
+	for j := 0; j < queues; j++ {
+		f.queues[j] = util.NewPriorityQueue(metric)
+	}
+
+	return f
+}
+
+// Enqueue adds the op to the next queue and prevents any other items to be added with this key
+func (f *ExclusiveQueues) Enqueue(op util.Op) {
+	_, ok := f.activeKeys.Load(op.Key())
+	if ok {
+		return
+	}
+
+	f.activeKeys.Store(op.Key(), struct{}{})
+	f.Requeue(op)
+}
+
+// Dequeue removes the next op from the requested queue.  After dequeueing the calling
+//  process either needs to call ClearKey or Requeue
+func (f *ExclusiveQueues) Dequeue(q int) util.Op {
+	return f.queues[q].Dequeue()
+}
+
+// Requeue adds an op that is presumed to already be covered by activeKeys
+func (f *ExclusiveQueues) Requeue(op util.Op) {
+	flushQueueIndex := int(f.index.Inc()) % len(f.queues)
+	f.queues[flushQueueIndex].Enqueue(op)
+}
+
+// ClearKey unblocks the requested key.  This should be called only after a flush has been successful
+func (f *ExclusiveQueues) Clear(op util.Op) {
+	f.activeKeys.Delete(op.Key())
```

Comment:
Wondering if we should delete the key on dequeue - there is no risk of double queueing because `enqueue` is protecting it. However, it would allow re-queueing in case it fails - is my understanding that with the clear happening after, the re-queueing will be a `noop` because the key still active.

Reply:
I think we do need to leave the key during dequeue.  After calling `Dequeue()` the calling code needs to either call `Requeue()` if there was a failure or `Clear()` if it succeeded.  Otherwise we could get a double queue during the period in which we are flushing.

The structure looks like this:
```
		op := i.flushQueues.Dequeue(j)

		err := i.flushUserTraces(op.userID)
		if err != nil {
			i.flushQueues.Requeue(op)
			continue
		}

		i.flushQueues.Clear(op)
```

Reply:
Yeah I saw that - so if it fails the requeue won't add the `opflush` because the key is still there.
```
	_, ok := f.activeKeys.Load(op.Key())
	if ok {
		return
	}
```

And because of the code above - no requeue is possible if the key already there.

Reply:
`Requeue()` doesn't check the map.  Only `Enqueue()` does.  `Requeue()` is designed to be used if you fail to handle the operation.  `Enqueue()` is used by the code generating the initial ops.

Reply:
I was being confused by the name `Enqueue` that wraps the actual queue. Thanks for clarifying.